cub/device/dispatch/dispatch_segmented_sort.cuh

```diff
@@ -0,0 +1,1434 @@
+   */
+  DoubleBuffer<KeyT> &d_keys;
+
+  /**
+   * Double-buffer whose current buffer contains the unsorted input values and,
+   * upon return, is updated to point to the sorted output values
+   */
+  DoubleBuffer<ValueT> &d_values;
+
+  /// Number of items to sort
+  OffsetT num_items;
+
+  /// The number of segments that comprise the sorting data
+  unsigned int num_segments;
+
+  /**
+   * Random-access input iterator to the sequence of beginning offsets of length
+   * @p num_segments, such that `d_begin_offsets[i]` is the first element of the
+   * <em>i</em><sup>th</sup> data segment in `d_keys_*` and `d_values_*`
+   */
+  BeginOffsetIteratorT d_begin_offsets;
+
+  /**
+   * Random-access input iterator to the sequence of ending offsets of length
+   * @p num_segments, such that <tt>d_end_offsets[i]-1</tt> is the last element
+   * of the <em>i</em><sup>th</sup> data segment in `d_keys_*` and
+   * `d_values_*`. If `d_end_offsets[i]-1 <= d_begin_offsets[i]`,
+   * the <em>i</em><sup>th</sup> is considered empty.
+   */
+  EndOffsetIteratorT d_end_offsets;
+
+  /// Whether is okay to overwrite source buffers
+  bool is_overwrite_okay;
+
+  /// CUDA stream to launch kernels within.
+  cudaStream_t stream;
+
+  /**
+   * Whether or not to synchronize the stream after every kernel launch to check
+   * for errors. Also causes launch configurations to be printed to the console.
+   */
+  bool debug_synchronous;
+
+  CUB_RUNTIME_FUNCTION __forceinline__
+  DispatchSegmentedSort(void *d_temp_storage,
+                        std::size_t &temp_storage_bytes,
+                        DoubleBuffer<KeyT> &d_keys,
+                        DoubleBuffer<ValueT> &d_values,
+                        OffsetT num_items,
+                        unsigned int num_segments,
+                        BeginOffsetIteratorT d_begin_offsets,
+                        EndOffsetIteratorT d_end_offsets,
+                        bool is_overwrite_okay,
+                        cudaStream_t stream,
+                        bool debug_synchronous)
+      : d_temp_storage(d_temp_storage)
+      , temp_storage_bytes(temp_storage_bytes)
+      , d_keys(d_keys)
+      , d_values(d_values)
+      , num_items(num_items)
+      , num_segments(num_segments)
+      , d_begin_offsets(d_begin_offsets)
+      , d_end_offsets(d_end_offsets)
+      , is_overwrite_okay(is_overwrite_okay)
+      , stream(stream)
+      , debug_synchronous(debug_synchronous)
+  {}
+
+  template <typename ActivePolicyT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t Invoke()
+  {
+    using LargeSegmentPolicyT = typename ActivePolicyT::LargeSegmentPolicy;
+    using SmallAndMediumPolicyT =
+      typename ActivePolicyT::SmallAndMediumSegmentedSortPolicyT;
+
+    constexpr int radix_bits = LargeSegmentPolicyT::RADIX_BITS;
+
+    cudaError error = cudaSuccess;
+
+    do
+    {
+      if (num_items == 0 || num_segments == 0)
+      {
+        temp_storage_bytes = 0;
```

Comment:
> Does this behave as expected, and do we have tests for this conditions?

We have `TestZeroSegments` and `TestEmptySegments` for this purpose. 

> IIRC, cudaMalloc(0) returns nullptr, which will throw off the d_temp_storage pointer check later on in the algorithm. 

This check doesn't depend on the `d_temp_storage`. Since the algorithm exits at this point, there won't be any later checks.